crates/bin/pcli/src/command/view/noble_address.rs

```diff
@@ -22,31 +36,243 @@ impl NobleAddressCmd {
         true
     }
 
-    pub fn exec(&self, fvk: &FullViewingKey) -> Result<()> {
-        let index: Result<u32, _> = self.address_or_index.parse();
-
-        let address = if let Ok(index) = index {
-            // address index provided
-            let (address, _dtk) = match self.ephemeral {
-                false => fvk.incoming().payment_address(index.into()),
-                true => fvk.incoming().ephemeral_address(OsRng, index.into()),
-            };
+    pub async fn exec(&self, fvk: &FullViewingKey) -> Result<()> {
+        let account = self.account;
+        let channel = &self.channel;
+        let noble_node = &self.noble_node;
 
-            address
-        } else {
-            // address or nothing provided
-            let address: Address = self
-                .address_or_index
-                .parse()
-                .map_err(|_| anyhow::anyhow!("Provided address is invalid."))?;
+        let next_sequence: u16 =
+            get_next_noble_sequence(Some(account), &fvk, &channel, &noble_node).await?;
 
-            address
-        };
+        let address = get_forwarding_address_for_sequence(next_sequence, Some(account), &fvk);
+        let noble_address = address.noble_forwarding_address(&channel);
 
-        let noble_address = address.noble_forwarding_address(&self.channel);
+        println!("next one-time use Noble forwarding address for account {} is: {}\n\nplease deposit funds to this address...\n\nawaiting deposit...\n\n", account, noble_address);
 
-        println!("{}", noble_address);
+        wait_for_noble_deposit(&noble_node, &noble_address, &address, &channel).await?;
+        println!(
+                    "💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫💫\n\nregistered Noble forwarding account with address {} to forward to Penumbra address {}...\n\nyour funds should show up in your Penumbra account shortly",
+                    noble_address, address
+                );
 
         Ok(())
     }
 }
+
+async fn get_next_noble_sequence(
+    account: Option<u32>,
+    fvk: &FullViewingKey,
+    channel: &str,
+    noble_node: &Url,
+) -> Result<u16> {
+    // perform binary search to find the first unused noble sequence number
+    // search space (sequence number) is 2 bytes wide
+    let left = 0u16;
+    let right = 0xffffu16;
+    let mid = (left + right) / 2u16;
+
+    // attempt to register midpoint
+    _get_next_noble_sequence(left, right, mid, noble_node, channel, fvk, account).await
+}
```

Comment:
Here is a web implementation of binary search, inspired by this, but handles edge cases a bit differently: https://github.com/prax-wallet/prax/pull/215/files#diff-c569e62334fe6b19c6a1434837ca111daa9f4eb8287e7ab0c4925564c2ff44c2